src/Storages/MergeTree/IntersectionsIndexes.h

```diff
@@ -0,0 +1,206 @@
+                if ((x.begin <= y.begin) && (y.begin <= x.end))
+                    return true;
+                if ((y.begin <= x.begin) && (x.begin <= y.end))
+                    return true;
+                return false;
+            };
+
+            if (are_intersect(segment.range, part.range))
+            {
+                /// We have two or possibly more intersections
+                if (intersected_before)
+                    return IntersectionResult::REJECT;
+
+                /// We have intersection with part with different name
+                /// It could happens if we have merged part on one replica
+                /// but not on another.
+                if (segment.name != part.name)
+                    return IntersectionResult::REJECT;
+
+                intersected_before = true;
+            }
+        }
+
+        return intersected_before ? IntersectionResult::EXACTLY_ONE_INTERSECTION : IntersectionResult::NO_INTERSECTION;
+    }
+
+    using OrderedSegments = std::set<PartToRead>;
+    OrderedSegments segments;
+};
+
+struct HalfIntervals
+{
+    static HalfIntervals initializeWithEntireSpace()
+    {
+        auto left_inf = std::numeric_limits<decltype(MarkRange::begin)>::min();
+        auto right_inf = std::numeric_limits<decltype(MarkRange::end)>::max();
+        return HalfIntervals{{{left_inf, right_inf}}};
+    }
+
+    static HalfIntervals initializeFromMarkRanges(MarkRanges ranges)
+    {
+        OrderedRanges new_intervals;
+        for (const auto & range : ranges)
+            new_intervals.insert(range);
+
+        return HalfIntervals{std::move(new_intervals)};
+    }
+
+    MarkRanges convertToMarkRangesFinal()
+    {
+        MarkRanges result;
+        std::move(intervals.begin(), intervals.end(), std::back_inserter(result));
+        return result;
+    }
+
+    HalfIntervals & intersect(const HalfIntervals & rhs)
+    {
+        /**
+         * first   [   ) [   ) [   ) [  ) [  )
+         * second    [       ) [ ) [   )  [    )
+         */
+        OrderedRanges intersected;
+
+        const auto & first_intervals = intervals;
+        auto first = first_intervals.begin();
+        const auto & second_intervals = rhs.intervals;
+        auto second = second_intervals.begin();
+
+        while (first != first_intervals.end() && second != second_intervals.end())
+        {
+            auto are_intersect = [](auto & x, auto & y)
+            {
+                if ((x->begin <= y->begin) && (y->begin < x->end))
+                    return true;
+                if ((y->begin <= x->begin) && (x->begin < y->end))
+                    return true;
+                return false;
+            };
+
+            if (are_intersect(first, second))
+            {
+                intersected.insert(MarkRange{
+                    std::max(second->begin, first->begin),
+                    std::min(second->end, first->end)
+                });
```

Comment:
btw, we can always calculate `range = MarkRange{std::max(second->begin, first->begin), std::min(second->end, first->end)}` but insert it only if `range.begin < range.end`. It will be correct intersection check.

Reply:
Ok